pkg/controllers/nodeclass/nodeclass_test.go

```diff
@@ -790,4 +792,106 @@ var _ = Describe("NodeClassController", func() {
 			Expect(nodeClass.Annotations[v1beta1.AnnotationNodeClassHash]).To(Equal(expectedHash))
 		})
 	})
+	Context("NodeClass Termination", func() {
+		var profileName string
+		BeforeEach(func() {
+			nodeClass.Spec.Role = "test-role"
+			profileName = awsEnv.InstanceProfileProvider.GetProfileName(ctx, nodeClass)
+		})
+		It("should succeed to delete the instance profile with no associated instances", func() {
+			awsEnv.IAMAPI.InstanceProfiles = map[string]*iam.InstanceProfile{
+				profileName: {
+					InstanceProfileName: aws.String(profileName),
+					Roles: []*iam.Role{
+						{
+							RoleId:   aws.String(fake.RoleID()),
+							RoleName: aws.String(nodeClass.Spec.Role),
+						},
+					},
+				},
+			}
+			ExpectApplied(ctx, env.Client, nodeClass)
+			ExpectReconcileSucceeded(ctx, nodeClassController, client.ObjectKeyFromObject(nodeClass))
+			Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(1))
+
+			Expect(env.Client.Delete(ctx, nodeClass)).To(Succeed())
+			ExpectReconcileSucceeded(ctx, nodeClassController, client.ObjectKeyFromObject(nodeClass))
+			Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(0))
+			ExpectNotFound(ctx, env.Client, nodeClass)
+		})
+		It("should succeed to delete the instance profile when no roles exist with no associated instances", func() {
+			awsEnv.IAMAPI.InstanceProfiles = map[string]*iam.InstanceProfile{
+				profileName: {
+					InstanceProfileName: aws.String(profileName),
+				},
+			}
+			ExpectApplied(ctx, env.Client, nodeClass)
+			ExpectReconcileSucceeded(ctx, nodeClassController, client.ObjectKeyFromObject(nodeClass))
+			Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(1))
+
+			Expect(env.Client.Delete(ctx, nodeClass)).To(Succeed())
+			ExpectReconcileSucceeded(ctx, nodeClassController, client.ObjectKeyFromObject(nodeClass))
+			Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(0))
+			ExpectNotFound(ctx, env.Client, nodeClass)
+		})
+		It("should succeed to delete the instance profile when it doesn't exist", func() {
+			ExpectApplied(ctx, env.Client, nodeClass)
+			ExpectReconcileSucceeded(ctx, nodeClassController, client.ObjectKeyFromObject(nodeClass))
+			Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(0))
+
+			Expect(env.Client.Delete(ctx, nodeClass)).To(Succeed())
+			ExpectReconcileSucceeded(ctx, nodeClassController, client.ObjectKeyFromObject(nodeClass))
+			Expect(awsEnv.IAMAPI.InstanceProfiles).To(HaveLen(0))
+			ExpectNotFound(ctx, env.Client, nodeClass)
+		})
+		It("should not delete the instance profile until all associated instances are terminated", func() {
+			var ids []string
+			for i := 0; i < 10; i++ {
```

Comment:
Probably easier to do just 2 instances here since we're just going to be doing the following, and how many instances over 1 doesn't matter. 
- Create n instances
- Try to delete
- Make it 1 instance left
- Try to delete
- Make 0 instances
- Delete successful